velox/functions/sparksql/aggregates/DecimalAverageAggregate.h

```diff
@@ -0,0 +1,245 @@
+
+    AccumulatorType() = delete;
+
+    AccumulatorType(
+        HashStringAllocator* /*allocator*/,
+        DecimalAverageAggregate* fn)
+        : fn(fn) {}
+
+    bool addInput(
+        HashStringAllocator* /*allocator*/,
+        exec::optional_arg_type<TInputType> data) {
+      if (!data.has_value()) {
+        return false;
+      }
+      if (!sum.has_value()) {
+        // sum is initialized to 0. When it is nullopt, it implies that the
+        // count number of the input values must not be 0.
+        VELOX_CHECK(count != 0);
+        return true;
+      }
+      int128_t result;
+      overflow +=
+          DecimalUtil::addWithOverflow(result, data.value(), sum.value());
+      sum = result;
+      count += 1;
+      return true;
+    }
+
+    bool combine(
+        HashStringAllocator* /*allocator*/,
+        exec::optional_arg_type<Row<int128_t, int64_t>> other) {
+      if (!other.has_value()) {
+        return false;
+      }
+      auto const otherSum = other.value().template at<0>();
+      auto const otherCount = other.value().template at<1>();
+
+      // otherCount is never null.
+      VELOX_CHECK(otherCount.has_value());
+      if (count == 0 && otherCount.value() == 0) {
+        // Both accumulators have no input values, no need to do the
+        // combination.
+        return false;
+      }
+
+      bool currentOverflow = count > 0 && !sum.has_value();
+      bool otherOverflow = otherCount.value() > 0 && !otherSum.has_value();
+      if (currentOverflow || otherOverflow) {
+        sum = std::nullopt;
+        count += otherCount.value();
+      } else {
+        int128_t result;
+        overflow +=
+            DecimalUtil::addWithOverflow(result, otherSum.value(), sum.value());
+        sum = result;
+        count += otherCount.value();
+      }
+      return true;
+    }
+
+    bool writeIntermediateResult(
+        bool nonNullGroup,
+        exec::out_type<IntermediateType>& out) {
+      if (!nonNullGroup) {
+        // If a group is null, all values in this group are null. In Spark, this
+        // group will be the initial value, where sum is 0 and count is 0.
+        out =
+            std::make_tuple(static_cast<int128_t>(0), static_cast<int64_t>(0));
+      } else {
+        if (!sum.has_value()) {
+          // Sum should be set to null on overflow.
+          out.template set_null_at<0>();
+          out.template get_writer_at<1>() = count;
+        } else {
+          auto adjustedSum =
+              DecimalUtil::adjustSumForOverflow(sum.value(), overflow);
+          if (adjustedSum.has_value()) {
+            out = std::make_tuple(adjustedSum.value(), count);
+          } else {
+            out.template set_null_at<0>();
+            out.template get_writer_at<1>() = count;
+          }
+        }
+      }
+      return true;
+    }
+
+    bool writeFinalResult(bool nonNullGroup, exec::out_type<OutputType>& out) {
+      if (!nonNullGroup || count == 0) {
```

Comment:
Does `count == 0` indicates that the result is null for empty input?

Reply:
Yes. This is consistent with https://github.com/apache/spark/blob/a8e35c407bc5340f83b35e5a2f0b0767c6baadb0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala#L107.

Reply:
Do we have a test to cover this?

Reply:
I added 2 test cases to verify scenarios where the input is empty or when the intermediate data count is 0. However, neither case can construct the situation where count equals 0 in `writeFinalResult`. This is because both `combine` and `writeIntermediateResult` will return false when count is 0 (resulting in a null group). The count in `writeFinalResult` can only potentially become 0 in the extract function, but Spark cannot register this companion function. Nevertheless, I believe keeping the `count == 0` check in `writeFinalResult` is still appropriate.